iam-login-service/src/main/java/it/infn/mw/iam/core/web/aup/AupReminderTask.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) Istituto Nazionale di Fisica Nucleare (INFN). 2016-2021
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package it.infn.mw.iam.core.web.aup;
+
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import it.infn.mw.iam.notification.NotificationFactory;
+import it.infn.mw.iam.persistence.model.IamAup;
+import it.infn.mw.iam.persistence.model.IamAupSignature;
+import it.infn.mw.iam.persistence.repository.IamAccountRepository;
+import it.infn.mw.iam.persistence.repository.IamAupRepository;
+import it.infn.mw.iam.persistence.repository.IamAupSignatureRepository;
+import it.infn.mw.iam.persistence.repository.IamEmailNotificationRepository;
+
+@Component
+public class AupReminderTask {
+
+  @Autowired
+  IamAccountRepository accounts;
+
+  @Autowired
+  IamAupRepository aupRepo;
+
+  @Autowired
+  NotificationFactory notification;
+
+  @Autowired
+  IamAupSignatureRepository aupSignatureRepo;
+
+  @Autowired
+  IamEmailNotificationRepository emailNotificationRepo;
+
+  public void sendAupReminders() {
+    aupRepo.findDefaultAup().ifPresent(aup -> {
+      LocalDate currentDate = LocalDate.now();
+      LocalDate expirationDate = currentDate.minusDays(aup.getSignatureValidityInDays());
+      Date expirationDateAsDate = toDate(expirationDate);
+      Date expirationDatePlusOneDayAsDate = toDate(expirationDate.plusDays(1));
+      List<Integer> reminderIntervals = parseReminderIntervals(aup.getAupRemindersInDays());
+
+      reminderIntervals.forEach(
+          interval -> processRemindersForInterval(aup, currentDate, interval, expirationDate));
+
+      List<IamAupSignature> expiredSignatures = aupSignatureRepo.findByAupAndSignatureTime(aup,
+          expirationDateAsDate, expirationDatePlusOneDayAsDate);
+
+      // check if an email of type AUP_EXPIRATION does not already exist, because it is never deleted
+      expiredSignatures.forEach(s -> {
+        if (emailNotificationRepo
```

Comment:
I'd prefer to define a method that hides how this check is implemented but explains with its name the real purpose: for example you could define a method called !expiredSignatureEmailAlreadySentFor(IamAccount a)

This makes code more readable